release/cibuild.py

```diff
@@ -10,163 +10,254 @@
-    else:
-        click.echo("Not building Docker image.")
-
-    if "PYINSTALLER" in os.environ:
-        build_pyinstaller()
-    else:
-        click.echo("Not building PyInstaller packages.")
-
+class BuildEnviron:
+    PLATFORM_TAGS = {
+        "Darwin": "osx",
+        "Windows": "windows",
+        "Linux": "linux",
+    }
+
+    def __init__(
+        self,
+        *,
+        system = "",
+        root_dir = "",
+
+        travis_tag = "",
+        travis_branch = "",
+        travis_pull_request = "",
+
+        appveyor_repo_tag_name = "",
+        appveyor_repo_branch = "",
+        appveyor_pull_request_number = "",
+
+        should_build_wheel = False,
+        should_build_docker = False,
+        should_build_pyinstaller = False,
+
+        has_aws_creds = False,
+        has_twine_creds = False,
+
+        docker_username = "",
+        docker_password = "",
+    ):
+        self.system = system
+        self.root_dir = root_dir
+
+        self.travis_tag = travis_tag
+        self.travis_branch = travis_branch
+        self.travis_pull_request = travis_pull_request
+
+        self.should_build_wheel = should_build_wheel
+        self.should_build_docker = should_build_docker
+        self.should_build_pyinstaller = should_build_pyinstaller
+
+        self.appveyor_repo_tag_name = appveyor_repo_tag_name
+        self.appveyor_repo_branch = appveyor_repo_branch
+        self.appveyor_pull_request_number = appveyor_pull_request_number
+
+        self.has_aws_creds = has_aws_creds
+        self.has_twine_creds = has_twine_creds
+        self.docker_username = docker_username
+        self.docker_password = docker_password
+
+    @classmethod
+    def from_env(klass):
+        return klass(
+            system = platform.system(),
+            root_dir = os.path.normpath(os.path.join(os.path.dirname(__file__), "..")),
+
+            travis_tag = os.environ.get("TRAVIS_TAG", ""),
+            travis_branch = os.environ.get("TRAVIS_BRANCH", ""),
+            travis_pull_request = os.environ.get("TRAVIS_PULL_REQUEST"),
+
+            appveyor_repo_tag_name = os.environ.get("APPVEYOR_REPO_TAG_NAME", ""),
+            appveyor_repo_branch = os.environ.get("APPVEYOR_REPO_BRANCH", ""),
+            appveyor_pull_request_number = os.environ.get("APPVEYOR_PULL_REQUEST_NUMBER"),
+
+            should_build_wheel = "WHEEL" in os.environ,
+            should_build_pyinstaller = "PYINSTALLER" in os.environ,
+            should_build_docker = "DOCKER" in os.environ,
+
+            has_aws_creds = "AWS_ACCESS_KEY_ID" in os.environ,
+            has_twine_creds= (
+                "TWINE_USERNAME" in os.environ and
+                "TWINE_PASSWORD" in os.environ
+            ),
+
+            docker_username = os.environ.get("DOCKER_USERNAME"),
+            docker_password = os.environ.get("DOCKER_PASSWORD"),
+        )
+
+    @property
+    def has_docker_creds(self) -> bool:
+        return self.docker_username and self.docker_password
```

Comment:
docker_creds and twine_creds are different.
`has_twine_creds` is an attribute, but `has_docker_creds` is a function-property... why?

Reply:
The docker creds have to be passed to docker login as arguments. The twine creds never leave the environment.

Reply:
👍